src/mbgl/map/source.cpp

```diff
@@ -603,4 +606,58 @@ void Source::dumpDebugLogs() const {
     }
 }
 
+FeatureResults Source::featuresAt(const PrecisionPoint point, const uint8_t radius, const TransformState& transform) const {
+    // figure out tile (bounded by source max zoom)
+    LatLng p = transform.pointToLatLng(point);
+
+    const double sine = std::sin(p.latitude * M_PI / 180);
+    const double x = p.longitude / 360 + 0.5;
+    double y = 0.5 - 0.25 * std::log((1 + sine) / (1 - sine)) / M_PI;
+
+    y = y < -1 ? -1 : y > 1 ? 1 : y;
```

Comment:
Added some comments in https://github.com/mapbox/mapbox-gl-native/commit/ac655cebe68065322eb773d5b96fa9dc80ddabc8, but see also https://github.com/mapbox/mapbox-gl-native/pull/3172#discussion_r47561929.